test/TritonGPU/combine.mlir

```diff
@@ -214,671 +214,16 @@ tt.func @if_else_both_convert(%arg0: i32, %arg1: !tt.ptr<i32> {tt.divisibility =
 #blocked1 = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [32, 1], warpsPerCTA = [4, 1], order = [0, 1], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>
 #slice1dim1 = #triton_gpu.slice<{dim = 1, parent = #blocked1}>
 #blocked2 = #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [1, 32], warpsPerCTA = [1, 4], order = [0, 1], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>
+#blocked2a = #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>
 #slice2dim0 = #triton_gpu.slice<{dim = 0, parent = #blocked2}>
 #blocked3 = #triton_gpu.blocked<{sizePerThread = [1, 4], threadsPerWarp = [2, 16], warpsPerCTA = [1, 4], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [1, 0]}>
 #blocked4 = #triton_gpu.blocked<{sizePerThread = [4, 1], threadsPerWarp = [16, 2], warpsPerCTA = [4, 1], order = [0, 1], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>
 #blocked5 = #triton_gpu.blocked<{sizePerThread = [4], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>
 
-// CHECK: [[$row_layout:#.*]] = #triton_gpu.blocked<{sizePerThread = [1, 4], threadsPerWarp = [2, 16], warpsPerCTA = [1, 4], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [1, 0]}>
```

Comment:
I'm concerned about removing those tests. Some of them recently caught problem for me. For instance the long one is meant to check that our algorithm doesn't have combinatory explosion. Those tests are very useful IMO

Reply:
These tests have completely invalid layouts, like tens of instructions are invalid IR.  I can "fix" them by choosing random layouts, although it's painstaking.  Do you want me to do that?  If so, how do I choose the "correct" layout?

Reply:
interesting, I think I need to understand what's invalid in those layouts

Reply:
Sure, happy to help.

I've pushed a branch with these tests added back.  So to reproduce, you can check out the PR, build, and then run

`$ triton-opt -split-input-file -tritongpu-remove-layout-conversions test/TritonGPU/combine.mlir > /dev/null`

It only prints a few of the errors, so it's an incremental process to understand what all is going wrong.  Here is the first batch of errors.

```
within split at test/TritonGPU/combine.mlir:211 offset :204:9: error: Result has an invalid layout: #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [2], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>.
Layout has a total of 2 warps per CTA, but the module specifies 4 warps per CTA.
  %14 = triton_gpu.convert_layout %13 : (tensor<256xf32, #blocked5>) -> tensor<256xf32, #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [2], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>
        ^
within split at test/TritonGPU/combine.mlir:211 offset :204:9: note: see current operation: %15 = "triton_gpu.convert_layout"(%14) : (tensor<256xf32, #triton_gpu.blocked<{sizePerThread = [4], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>) -> tensor<256xf32, #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [2], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>
within split at test/TritonGPU/combine.mlir:1573 offset :44:11: error: Result has an invalid layout: #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>.
Layout has rank 2, but the tensor it's attached to has rank 1.
    %18 = tt.splat %7 : (i64) -> tensor<128xi64, #blocked3>
          ^
within split at test/TritonGPU/combine.mlir:1573 offset :44:11: note: see current operation: %28 = "tt.splat"(%17) : (i64) -> tensor<128xi64, #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>>
```

You can follow the IR to see these invalid layouts continuing to be used in the tests.  If I delete these functions, I get more brokenness.  For example, after deleting `@vecadd`, I get

```
within split at test/TritonGPU/combine.mlir:211 offset :268:8: error: Result has an invalid layout: #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [32, 1], warpsPerCTA = [4, 1], order = [0, 1], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>.
Layout has rank 2, but the tensor it's attached to has rank 1.
  %8 = triton_gpu.convert_layout %7 : (tensor<1024x!tt.ptr<f32>, #blocked0>) -> tensor<1024x!tt.ptr<f32>, #blocked1>
       ^
within split at test/TritonGPU/combine.mlir:211 offset :268:8: note: see current operation: %26 = "triton_gpu.convert_layout"(%25) : (tensor<1024x!tt.ptr<f32, 1>, #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>) -> tensor<1024x!tt.ptr<f32, 1>, #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [32, 1], warpsPerCTA = [4, 1], order = [0, 1], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>>
within split at test/TritonGPU/combine.mlir:1540 offset :44:11: error: Result has an invalid layout: #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>.
Layout has rank 2, but the tensor it's attached to has rank 1.
    %18 = tt.splat %7 : (i64) -> tensor<128xi64, #blocked3>
          ^
within split at test/TritonGPU/combine.mlir:1540 offset :44:11: note: see current operation: %28 = "tt.splat"(%17) : (i64) -> tensor<128xi64, #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>>
```

and then after deleting `long_func`, I get

```
within split at test/TritonGPU/combine.mlir:211 offset :204:9: error: Result has an invalid layout: #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [2], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>.
Layout has a total of 2 warps per CTA, but the module specifies 4 warps per CTA.
  %14 = triton_gpu.convert_layout %13 : (tensor<256xf32, #blocked5>) -> tensor<256xf32, #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [2], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>
        ^
within split at test/TritonGPU/combine.mlir:211 offset :204:9: note: see current operation: %15 = "triton_gpu.convert_layout"(%14) : (tensor<256xf32, #triton_gpu.blocked<{sizePerThread = [4], threadsPerWarp = [32], warpsPerCTA = [4], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>) -> tensor<256xf32, #triton_gpu.blocked<{sizePerThread = [1], threadsPerWarp = [32], warpsPerCTA = [2], order = [0], CTAsPerCGA = [1], CTASplitNum = [1], CTAOrder = [0]}>>
within split at test/TritonGPU/combine.mlir:1175 offset :44:11: error: Result has an invalid layout: #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>.
Layout has rank 2, but the tensor it's attached to has rank 1.
    %18 = tt.splat %7 : (i64) -> tensor<128xi64, #blocked3>
          ^
within split at test/TritonGPU/combine.mlir:1175 offset :44:11: note: see current operation: %28 = "tt.splat"(%17) : (i64) -> tensor<128xi64, #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>>
```

All of these functions do seem incorrect.

I will try to fix them manually, but let me know if you think it's not worth it.

Reply:
Fixed this manually.